pkg/unifiedpush/client.go

```diff
@@ -0,0 +1,197 @@
+package unifiedpush
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	pushv1alpha1 "github.com/aerogear/unifiedpush-operator/pkg/apis/push/v1alpha1"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"k8s.io/apimachinery/pkg/runtime"
+	"mime/multipart"
+	"net/http"
+	"strconv"
+)
+
+// variant is an internal base type
+type variant struct {
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	VariantId   string `json:"variantID"`
+	Secret      string `json:"secret"`
+}
+
+// upsAndroidVariant is an internal struct used for convenient JSON
+// unmarshalling of the response received from UPS
+type upsAndroidVariant struct {
+	ProjectNumber string `json:"projectNumber"`
+	GoogleKey     string `json:"googleKey"`
+	variant
+}
+
+// upsIOSVariant is an internal struct used for convenient JSON
+// unmarshalling of the response received from UPS
+type upsIOSVariant struct {
+	Certificate []byte `json:"certificate"`
+	Passphrase  string `json:"passphrase"`
+	Production  bool   `json:"production"`
+	variant
+}
+
+// pushApplication is an internal struct used for convenient JSON
+// unmarshalling of the response received from UPS
+type pushApplication struct {
+	ApplicationId string `json:applicationId`
+	MasterSecret  string `json:masterSecret`
+}
+
+// UnifiedpushClient is a client to enable easy interaction with a UPS
+// server
+type UnifiedpushClient struct {
+	url string
+}
+
+// CreateApplication creates an application in UPS
+func (c UnifiedpushClient) CreateApplication(p *pushv1alpha1.PushApplication) (string, string, error) {
```

Comment:
IMO, as a convention, it is better to return structs instead of strings.
Passing pointer to the method and making rest client modifying a CR struct kinda sounds strange.
I am not the Golang authority though :)

Reply:
On a second though, when you want to return structs, you need to create new struct types just for the return value. In that case, probably the passing a pointer to the CR and modifying that is good.

Reply:
My final idea: leave it as is! :grin: 

Let's see where it brings us.